Gem install bug fixes

defect opened the pull request:
This pull request fixes/changes two things:

1) When installing gems using the DSL, include the gem sources specified in the config. Right now we can only install gems coming from rubygems.

2) `install :gem` tries to automagically require the gems we just installed. However, the name of the gem and the name we need to require isn't always the same which makes the task fail. This changes it so that we print a message if we fail to require the gem, and lets the user require it in the task. I would like to nuke the autoloading completely, but i also don't want to break backwards compatibility :)

@roymarantz @maddalab @byxorna @Primer42

byxorna commented on src/framework/lib/genesisframework/task.rb:
```diff
@@ -136,8 +138,14 @@ def install provider, *what
+              begin
+                require gem
+              rescue LoadError
+                puts "Could not load gem #{gem} automatically, please load it explicitly in the task"
```
Add a thing about "maybe the gem name is different from its require path?"

byxorna commented:
LGTM, save for the more descriptive error message :)

Primer42 commented on src/framework/lib/genesisframework/task.rb:
```diff
@@ -124,7 +124,9 @@ def install provider, *what
             end
             if gems.size > 0    # make sure we still have something to do
-              Kernel.system('gem', 'install', '--no-ri', '--no-rdoc', *gems)
+              options = ['--no-ri', '--no-rdoc', config.fetch(:gems_source, '').split]
```
Doesn't this need a `--clear-sources` and some `--source` flags? Or am I misunderstanding the value that will be in `:gem_source`?

defect replied:
In our setup we have something along the lines of `:gems_source: --clear-sources --source https://internal-gemserver.com` in the genesis config.

I'll add a `:gem_source:` to the example config in the testenv with some documentation.

Primer42 replied:
In that case, can we name the key something like `:gem_args` or `:gem_flags`, since this could potentially be used to add other flags to the gem install command.

roymarantz commented:
I agree that doing a hidden require is a little smelly, but it is handy.

An alternative to breaking that and fixing all tasks would be to extend the syntax of install :gem such that the arguments can be a list of [gem, require] name pairs instead of a list of gem names only.  i.e.
install :gem, ['gem1', 'require1'], :gem2, ....
or if you prefer a hash
install :gem 'gem1' => 'require1', 'gem2' => 'gem2', ...

obviously we would need to be careful to parse:
install :gem, 'gem1', 'gem2'
correctly as install gem1 gem1 ; require gem1 gem2
and not as install gem1; require gem2

Bother these have there own amoung of smell too.

defect commented:
@roymarantz I've added functionality to specify what to require to the `install :gem` DSL method now. This:
```
install :gem, 'gem1', 'gem2' => ['gem2/foo', 'gem2/bar']
```
will install `gem1` and `gem2` and require `gem1`, `gem2/foo` and
`gem2/bar`.

I've also changed the name of `gems_source` to `gem_args` and added an example to the testenv example config.

roymarantz commented on src/framework/lib/genesisframework/task.rb:
```diff
@@ -112,32 +112,55 @@ def install provider, *what
+              if item.is_a?(Hash)
+                gems.merge! item
+              else
+                gems[item] = []
```
I think it is slightly better to do:
gem[item] = [item]
instead to keep all the complexity around standardizing the given to gem_name... structure local to this block.
Not a big deal, but a slight win IMO.

defect replied:
Oh yeah, totally. Nice one! Thanks! Commit incoming in a minute.

roymarantz commented on src/framework/lib/genesisframework/task.rb:
```diff
@@ -112,32 +112,55 @@ def install provider, *what
+            # Attempt to require loaded gems, print a message if we can't.
+            gems.each { |gem, requires|
+              begin
+                if requires.empty?
```
If you do my suggestion above, then this code become much simpler.  YMMV

roymarantz commented on testenv/bootbox/puppet/modules/genesis/templates/config.yaml.erb:
```diff
@@ -19,6 +19,7 @@
         :gpgcheck:  0
 :gem_files:
     "genesis_framework": http://<%= @genesis_ipaddress %>:8888/gem/genesis_framework
+:gem_args: &gem_args --no-ri --no-rdoc --clear-sources # Arguments to pass to gem install. Can include --source for internal repos
```
I don't think you should include --clear-sources here since without a --source the gem install will probably fail.  So either:
:gem_args: &gem_args --no-ri --no-rdoc
or
:gem_args: &gem_args --no-ri --no-rdoc --clear-sources --source https://gems.example.com

roymarantz commented:
I'm fine with this once the sample gem_args is fixed although I'd prefer the code change too.
I really like this.  Good work. :100:

defect commented:
Thanks @roymarantz. Makes a lot of sense. I've fixed your comments, now go celebrate july 4th! :P

roymarantz commented on src/framework/lib/genesisframework/task.rb:
```diff
@@ -112,32 +112,51 @@ def install provider, *what
+              begin
+                requires.each {|r| require r }
+              rescue LoadError
+                puts "Could not load gem #{gem} automatically. Maybe the gem name differs from its load path? Please specify the name to require, or load it explicitly in the task."
```
Why is this a #puts instead of #raise ?  You don't handle/fix the error in any way so another error will just happen later and this message will probably be lost in the confusion.  You may want to #log this message too so it is visible in the asset's collins log.

defect replied:
The idea was to let people require the gem themselves in the task if we failed to load it automatically, but I guess now that we provide a way to specify requires with `install :gem` we probably should raise an error as you say.

roymarantz replied:
I believe that the install :gem would fail and abort the task if the require didn't work.  Rescuing the error and just printing a message would allow the task to continue leading to a problem later when the gem was accessed.  I don't think that should happen hence the raise of your better error message. 

Logging the message just makes the message (potentially) easier to find and documents why the task (provisioning) failed even if the node rebooted.

defect commented:
This has been running happily in production in our environment for about a week now. Unless anyone objects I'll merge this on monday.
